python/tk_nuke/context.py

```diff
@@ -0,0 +1,373 @@
+            return
+
+        # Set the current context to be remembered for the next context
+        # change.
+        self._current_studio_context = focusInNuke
+
+        if focusInNuke:
+            # We switched from the project timeline to a Nuke node graph.
+            try:
+                script_path = nuke.scriptName()
+            except Exception:
+                script_path = None
+
+            if script_path:
+                # Switched to nuke with a script open. We have a path and could try
+                # to figure out the sgtk context from that.
+                new_context = self._get_new_context(script_path)
+
+                if new_context is not None and new_context != self.engine.context:
+                    self._change_context(new_context)
+            else:
+                # There is no script open in the nuke node graph. Empty
+                # session so we won't switch to a new context, although we
+                # should rebuild the menu to reflect that.
+                self.engine.menu_generator.create_sgtk_disabled_menu()
+        else:
+            # This is a switch back to the project-level timeline,
+            # so change to that context based on that project file's
+            # path.
+            project_path = self._get_current_project()
+            if project_path is not None:
+                self._change_context(self._get_new_context(project_path))
+
+    def _get_context_from_script(self, script):
+        """
+        Returns an sgtk.context.Context object from the given script path.
+
+        :param script:  The path to a script file on disk.
+        """
+        tk = tank.tank_from_path(script)
+
+        context = tk.context_from_path(
+            script,
+            previous_context=self.engine.context,
+        )
+
+        if context.project is None:
+            raise tank.TankError(
+                "The Nuke engine needs at least a project "
+                "context in order to start! Your context: %s" % context
+            )
+        else:
+            return context
+
+    def _get_current_project(self):
+        """
+        Returns the current project based on where in the UI the user clicked.
+        """
+        import hiero.core
+        import hiero.ui
+
+        view = hiero.ui.activeView()
+        if isinstance(view, hiero.ui.TimelineEditor):
+            sequence = view.sequence()
+            project = sequence.binItem().project()
+            return project.path()
+        else:
+            return None
+
+    def _get_new_context(self, script_path):
+        """
+        Returns a new sgtk.context.Context for the given script path.
+
+        If the context exists in the in-memory cache, then that is returned,
+        otherwise a new Context object is constructed, cached, and returned.
+
+        :param script_path: The path to a script file on disk.
+        """
+        context = self._context_cache.get(script_path)
```

Comment:
Everytime I see this pattern, I wish we were using [Python 3](https://docs.python.org/3/library/functools.html#functools.lru_cache)
Or what we had implemented that pattern as a decorator in Toolkit.